src/analysis/modal-utils.ts

```diff
@@ -33,61 +39,136 @@ export interface IComputeConfig {
+
+export interface ExistingModalPersistentDiskConfig {
+  size: number;
+  diskType: GooglePdType;
+}
+// The desired diverges from the existing because the api request to make a runtime requires `GoogleDiskType`
+// However, we decorate the type to `GooglePdType` when we get it from the api
+export interface DesiredModalPersistentDiskConfig {
+  size: number;
+  diskType: GoogleDiskType;
+}
+
+export type NormalizedModalRuntime = {
+  cloudService: ComputeType | undefined;
+  toolDockerImage: string | undefined; // this is a url
+  tool?: ToolLabel;
+  persistentDiskAttached: boolean;
+  region: string;
+  jupyterUserScriptUri?: string;
+  gpuConfig?: any; // TODO
+  zone?: string;
+  machineType?: string;
+  bootDiskSize?: number | null;
+  diskSize?: number;
+  componentGatewayEnabled?: boolean;
+  masterMachineType?: string;
+  masterDiskSize?: number;
+  numberOfWorkers?: number;
+  numberOfPreemptibleWorkers?: number | null;
+  workerMachineType?: string | null | undefined;
+  workerDiskSize?: number | null | undefined;
+  timeoutInMinutes?: number | null;
+};
+
+// Auxiliary functions -- begin
+export const isGce = (runtimeType) => runtimeType === runtimeTypes.gceVm;
+
+export const isDataproc = (runtimeType) =>
+  runtimeType === runtimeTypes.dataprocSingleNode || runtimeType === runtimeTypes.dataprocCluster;
+
+export const isDataprocCluster = (runtimeType) => runtimeType === runtimeTypes.dataprocCluster;
+
+// TODO: use type
+export const runtimeTypes = {
+  gceVm: 'Standard VM',
+  dataprocSingleNode: 'Spark single node',
+  dataprocCluster: 'Spark cluster',
+};
+
+export const shouldUsePersistentDisk = (runtimeType, runtimeDetails, upgradeDiskSelected) =>
+  isGce(runtimeType) && (!runtimeDetails?.runtimeConfig?.diskSize || upgradeDiskSelected);
+
+// TODO: tnis is the best place to start
 export const buildExistingEnvironmentConfig = (
   computeConfig: IComputeConfig,
-  currentRuntimeDetails: any,
-  currentPersistentDiskDetails: PersistentDisk,
-  isDataproc: boolean
-) => {
+  currentRuntimeDetails?: GetRuntimeItem,
+  currentPersistentDiskDetails?: PersistentDisk,
+  isDataproc = false
+): ExistingModalRuntimeConfig => {
   const runtimeConfig = currentRuntimeDetails?.runtimeConfig;
   const cloudService = runtimeConfig?.cloudService;
-  const numberOfWorkers = runtimeConfig?.numberOfWorkers || 0;
-  const hasGpu = computeConfig.hasGpu;
-  const gpuConfig = runtimeConfig?.gpuConfig;
-  const toolLabel = getToolLabelFromCloudEnv(currentRuntimeDetails);
+  const numberOfWorkers = runtimeConfig && 'numberOfWorkers' in runtimeConfig ? runtimeConfig.numberOfWorkers : 0;
+  const gpuConfig = runtimeConfig && 'gpuConfig' in runtimeConfig ? runtimeConfig.gpuConfig : undefined;
+  const toolLabel = currentRuntimeDetails ? getToolLabelFromCloudEnv(currentRuntimeDetails) : undefined;
   return {
-    hasGpu,
+    hasGpu: computeConfig.hasGpu,
     autopauseThreshold: currentRuntimeDetails?.autopauseThreshold,
+    computeConfig,
```

Comment:
Should these two be typed as well?

Reply:
these are typed here https://github.com/DataBiosphere/terra-ui/pull/5079/files#diff-32a05189ab14623d039151b28dfee24908fc62cad51bf92afdb7767afbf7ef01R45